### PR TITLE
Dockerfile to serve this repo. 

### PR DESCRIPTION
**Pull request by snim2**
Build with 'docker build .' and run with something like 'sudo docker run -ti a16d7681862b -p 4000:4000'

Still needs more work, and some documentation in the README

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    && gem install \
+        bundler \
+        html-proofer \
+        jekyll \
+        jekyll-redirect-from \
+        kramdown \
+        rdiscount \
+        rouge \
+        rvm \
+    && wget https://github.com/PyconUK/pyconuk.github.io/archive/master.zip \
```
Comment (ghickman) [inline Dockerfile]: We shouldn't need to pull down the code base here, it should be running in the cloned repo already. I think this route would also cause issues with making local changes.

Comment (snim2) [inline Dockerfile]: +1. I'm writing a fix for this, but since the container takes an age to build on my machine it'll be a while until I push the commit.

**Comment by ghickman** [comment]
@snim2 – I was using [this Dockerfile](https://github.com/pypa/warehouse/blob/master/Dockerfile) recently and it has some directives in there that might be of use, specifically the `COPY` and `WORKDIR` ones?

We'll also need to test this out on OS X (where Docker isn't as simple) which I can hopefully look into tonight/this weekend.

**Comment by snim2** [comment]
Interesting, I was planning to use `ADD`.

**Comment by ghickman** [comment]
@snim2 – To be super clear here - I've not set up Docker before so please don't think I'm telling you what to use!

**Comment by snim2** [comment]
OK, pushed some plausible changes, but likely won't have time to properly test this until next week.

**Comment by inglesp** [comment]
I can build this on my OSX laptop, but when I try to run it, I can't connect to port 4000.

Can somebody else on OSX (@doismellburning, @ghickman?) give this a go?

**Comment by doismellburning** [comment]
@inglesp Are you sure you're connecting to the right IP? `boot2docker` uses a Vagrant box to run the Docker daemon - you don't want `localhost:4000` you want `$vagrantip:4000` - check `$DOCKER_HOST`

**Comment by ghickman** [comment]
@inglesp – `boot2docker ip` will give you the external IP of the docker instance. I couldn't get it to resolve under any of `localhost:4000`, `dockerip:4000` of `dockerip`. I tried telling jekyll to serve on the external IP by adding `-H 0.0.0.0` to the `ENTRYPOINT` section but it didn't seem to do anything.

**Comment by doismellburning** [comment]
@inglesp Oh sorry, I should have checked the `Dockerfile` - [it's still missing `--host 0.0.0.0`](https://github.com/PyconUK/pyconuk.github.io/issues/22#issuecomment-93993158)

**Comment by doismellburning** [comment]
So I lifted https://gist.github.com/doismellburning/60c97b2e688b92e6eba3 from another project, which may be of interest? It uses a few more intermediary layers for caching goodness, and doesn't duplicate the `Gemfile`

**Comment by doismellburning** [comment]
(Want me to PR it this way?)

**Comment by doismellburning** [comment]
(Also don't forget the port forward; try something like `docker build -t pyconukorg . && docker run -p 4000:4000 -ti pyconukorg`)

**Comment by ghickman** [comment]
I've tested out @doismellburning's Dockerfile locally and it all seems to be working for me.

One thing to note - we could serve the site at port 80 by changing the port mapping to `80:4000`.

**Comment by inglesp** [comment]
I've tested this with @doismellburning's Dockerfile, and it half works... I can run jekyll inside docker which builds the site, and I can access the site from my browser, but any changes to I make to the code are not reflected when I refresh the site until I restart the docker container.

In order to fix this, I've updated the Dockerfile (see https://gist.github.com/inglesp/08e5113f20124de0f3f2).  You'll then need to run `docker run -p 80:4000 -v $(pwd):/opt/pyconuk/site:rw -ti pyconuk.org`.

**Comment by doismellburning** [comment]
I'm terribly sorry for ever starting the trend of contribution-via-gist

**Comment by inglesp** [comment]
Meta question: what's the right way :tm: to do this kind of collaboration?

**Comment by doismellburning** [comment]
TBH probably multiple PRs - I was concerned it might seem rude though if I rocked up with my own totally distinct PR, but actually, turning up with a gist was probably worse, sorry :s

**Comment by snim2** [comment]
So, do you just want to close this PR and start another one?

**Comment by ntoll** [comment]
Hi,

I've been trying to get this to work by following the instructions at #26. Several observations:
- Couldn't we have just changed the README on this request rather than creating a new branch..?
- I suggest that people just update this branch rather than quote all those gists..! AFAICT it's a branch within the pyconuk organisation so we all have write access. Also, it's helpful to see the history of what's changed as the PR is evaluated. IMNSHO :-P
- So @ghickman's instructions over on #26 don't work. I can't connect to port 4000 (I'm on Debain 7 stable).

:-(

I propose we all:
- Fix stuff on this branch (since we all can)
- Use me as a technical crash test dummy
- Merge when it goes from zero -> working.

File: README.rst [hunk 2]
```diff
@@ -25,20 +25,25 @@ That's it!
 Usage
 -----
 
-This site uses Jekyll_. The official docs are good, but here's a quickstart:
+Installation
+~~~~~~~~~~~~
+This site uses Jekyll_ and we've wrapped the development environment in Docker.
```
Comment (inglesp) [inline README.rst]: Could we keep the jekyll instructions, and give people a choice to use jekyll or docker?

Comment (ghickman) [inline README.rst]: Good point, I'll restore and split into two sections.

File: README.rst [hunk 3]
```diff
@@ -33,6 +35,18 @@ This site uses Jekyll_. The official docs are good, but here's a quickstart:
 * ``bundle exec jekyll serve -w`` will run a development server listening on port 4000
 * ``bundle exec jekyll build`` will build the site in ``$gitroot/_site/``
 
+You can also set up a development environment with Docker:
+
+0. (OS X only) Install boot2docker_ (you can also homebrew it)
+1. Install docker_
+2. Build the docker image (this may take some time): ``make build``
+3. Run the docker image: ``make run``
+4. Get your [docker] ip. On OS X do: ``boot2docker ip``, one Linux this is your external IP
```
Comment (ntoll) [inline README.rst]: s/one Linux/on Linux

File: README.rst [hunk 4]
```diff
@@ -33,6 +35,18 @@ This site uses Jekyll_. The official docs are good, but here's a quickstart:
...
+5. Open your web browser of choice and go to the IP from step 4
```
Comment (ntoll) [inline README.rst]: Mention port 80... the 4000 thing is a tad confusing. ;-)

**Pull request merged**